### PR TITLE
Don't activate triggered events if the flow scope has been interrupted

### LINKED ISSUE #8726: Triggered event subprocess not activated after terminating boundary event
**Describe the bug**

Assume the following process:
image
If `task1` is being processed and we send a message, triggering the boundary event, the subprocess will be terminated. If we send a message immediately after the first message, triggering the interrupting event subprocess, while the subprocess is being terminated this will result in the event sub process not getting activated. This means the process instance will be stuck.

Note: This also happens when the boundary event is attached to a task without a subprocess. However, since there is less processing to be done when the boundary event gets triggered it is way less likely to happen.

**To Reproduce**

<details><summary>Unit test to reproduce</summary></p>

```java
  @Test
  public void reproduce() {
    final String processId = "process";
    final var process =
        Bpmn.createExecutableProcess(processId)
            .eventSubProcess(
                "eventSP",
                s ->
                    s.startEvent("eventSPStart")
                        .interrupting(true)
                        .message(
                            m -> m.name("eventMessage").zeebeCorrelationKey("=\"eventMessage\""))
                        .endEvent())
            .startEvent("start")
            .subProcess("sp")
            .embeddedSubProcess()
            .startEvent("spStart")
            .serviceTask("task1", t -> t.zeebeJobType("t1"))
            .endEvent()
            .subProcessDone()
            .boundaryEvent(
                "boundary",
                b -> b.message(m -> m.name("bMessage").zeebeCorrelationKey("=\"bMessage\"")))
            .serviceTask("task2", t -> t.zeebeJobType("t2"))
            .endEvent()
            .moveToActivity("sp")
            .endEvent()
            .done();

    ENGINE.deployment().withXmlResource(process).deploy();
    ENGINE.processInstance().ofBpmnProcessId(processId).create();
    RecordingExporter.jobRecords().withElementId("task1").await();
    ENGINE.message().withName("bMessage").withCorrelationKey("bMessage").publish();
    ENGINE.message().withName("eventMessage").withCorrelationKey("eventMessage").publish();

    assertThat(
            RecordingExporter.processInstanceRecords()
                .withElementType(BpmnElementType.END_EVENT)
                .exists())
        .isTrue();
  }
```

</p></details>

**Expected behavior**

I would expect the boundary event to get terminated (along with task1 and the subprocess) once the event subprocess gets triggered. The event sub process should then complete as expected.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
C DEPLOYMENT         CREATE         - #01-> -1  -1 - 
E PROC               CREATED        - #02->#01 K01 - process.xml -> "process" (version:1)
E DEPLOYMENT         CREATED        - #03->#01 K02 - process.xml
E DEPLOYMENT         FULLY_DISTR    - #04->#01 K02 - 
C PROC_INST_CREATION CREATE         - #05-> -1  -1 - new <process "process"> (no vars)
C PROC_INST          ACTIVATE       - #06->#05 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_INST_CREATION CREATED        - #07->#05 K04 - new <process "process"> (no vars)
E PROC_INST          ACTIVATING     - #08->#06 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_MSG_SUB       CREATING       - #09->#06 K05 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATED      - #10->#06 K03 - PROCESS "process" in <process "process"[K03]>
C PROC_INST          ACTIVATE       - #11->#06  -1 - START_EVENT "start" in <process "process"[K03]>
C MSG_SUB            CREATE         - #12-> -1  -1 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATING     - #13->#11 K06 - START_EVENT "start" in <process "process"[K03]>
E PROC_INST          ACTIVATED      - #14->#11 K06 - START_EVENT "start" in <process "process"[K03]>
C PROC_INST          COMPLETE       - #15->#11 K06 - START_EVENT "start" in <process "process"[K03]>
E MSG_SUB            CREATED        - #16->#12 K07 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
C PROC_MSG_SUB       CREATE         - #17-> -1  -1 - "eventMessage" (inter.) @[K03] in <process ?[K03]> (no vars)
E PROC_INST          COMPLETING     - #18->#15 K06 - START_EVENT "start" in <process "process"[K03]>
E PROC_INST          COMPLETED      - #19->#15 K06 - START_EVENT "start" in <process "process"[K03]>
E PROC_INST          SEQ_FLOW_TAKEN - #20->#15 K08 - SEQUENCE_FLOW "sequenc..64b0258" in <process "process"[K03]>
C PROC_INST          ACTIVATE       - #21->#15 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
E PROC_MSG_SUB       CREATED        - #22->#17 K05 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATING     - #23->#21 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
E PROC_MSG_SUB       CREATING       - #24->#21 K10 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATED      - #25->#21 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
C PROC_INST          ACTIVATE       - #26->#21  -1 - START_EVENT "spStart" in <process "process"[K03]>
C MSG_SUB            CREATE         - #27-> -1  -1 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATING     - #28->#26 K11 - START_EVENT "spStart" in <process "process"[K03]>
E PROC_INST          ACTIVATED      - #29->#26 K11 - START_EVENT "spStart" in <process "process"[K03]>
C PROC_INST          COMPLETE       - #30->#26 K11 - START_EVENT "spStart" in <process "process"[K03]>
E MSG_SUB            CREATED        - #31->#27 K12 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
C PROC_MSG_SUB       CREATE         - #32-> -1  -1 - "bMessage" (inter.) @[K09] in <process ?[K03]> (no vars)
E PROC_INST          COMPLETING     - #33->#30 K11 - START_EVENT "spStart" in <process "process"[K03]>
E PROC_INST          COMPLETED      - #34->#30 K11 - START_EVENT "spStart" in <process "process"[K03]>
E PROC_INST          SEQ_FLOW_TAKEN - #35->#30 K13 - SEQUENCE_FLOW "sequenc..5e9078f" in <process "process"[K03]>
C PROC_INST          ACTIVATE       - #36->#30 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
E PROC_MSG_SUB       CREATED        - #37->#32 K10 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATING     - #38->#36 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
E JOB                CREATED        - #39->#36 K15 - K15 "t1" @"task1"[K14] 3 retries, in <process "process"[K03]> (no vars)
E PROC_INST          ACTIVATED      - #40->#36 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
C MSG                PUBLISH        - #41-> -1  -1 - "bMessage" correlationKey: bMessage (no vars)
E MSG                PUBLISHED      - #42->#41 K16 - "bMessage" correlationKey: bMessage (no vars)
E MSG_SUB            CORRELATING    - #43->#41 K12 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
C PROC_MSG_SUB       CORRELATE      - #44-> -1  -1 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
C MSG                PUBLISH        - #45-> -1  -1 - "eventMessage" correlationKey: eventMessage (no vars)
E PROC_MSG_SUB       CORRELATED     - #46->#44 K10 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
E PROC_EVNT          TRIGGERING     - #47->#44 K17 -  @"boundary"[K09] in <process K01[K03]> (no vars)
C PROC_INST          TERMINATE      - #48->#44 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
C MSG_SUB            CORRELATE      - #49-> -1  -1 - "bMessage" (inter.) @[K09] in <process "process"[K03]> (no vars)
E MSG                PUBLISHED      - #50->#45 K18 - "eventMessage" correlationKey: eventMessage (no vars)
E MSG_SUB            CORRELATING    - #51->#45 K07 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
C PROC_MSG_SUB       CORRELATE      - #52-> -1  -1 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E PROC_INST          TERMINATING    - #53->#48 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
C PROC_INST          TERMINATE      - #54->#48 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
E MSG_SUB            CORRELATED     - #55->#49 K12 - "bMessage" (inter.) correlationKey: bMessage @[K09] in <process "process"[K03]> (no vars)
E PROC_MSG_SUB       CORRELATED     - #56->#52 K05 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E PROC_EVNT          TRIGGERING     - #57->#52 K19 -  @"eventSPStart"[K03] in <process K01[K03]> (no vars)
C MSG_SUB            CORRELATE      - #58-> -1  -1 - "eventMessage" (inter.) @[K03] in <process "process"[K03]> (no vars)
E PROC_INST          TERMINATING    - #59->#54 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
C JOB                CANCEL         - #60->#54 K15 - K15 "t1" @"task1"[K14] 3 retries, in <process "process"[K03]> (no vars)
E PROC_INST          TERMINATED     - #61->#54 K14 - SERVICE_TASK "task1" in <process "process"[K03]>
E PROC_INST          TERMINATED     - #62->#54 K09 - SUB_PROCESS "sp" in <process "process"[K03]>
E PROC_EVNT          TRIGGERED      - #63->#54 K17 -  @"boundary"[K09] in <process K01[K03]> (no vars)
E PROC_INST          ACTIVATING     - #64->#54 K20 - BOUNDARY_EVENT "boundary" in <process "process"[K03]>
E PROC_INST          ACTIVATED      - #65->#54 K20 - BOUNDARY_EVENT "boundary" in <process "process"[K03]>
C PROC_INST          COMPLETE       - #66->#54 K20 - BOUNDARY_EVENT "boundary" in <process "process"[K03]>
E MSG_SUB            CORRELATED     - #67->#58 K07 - "eventMessage" (inter.) correlationKey: eventMessage @[K03] in <process "process"[K03]> (no vars)
E JOB                CANCELED       - #68->#60 K15 - K15 "t1" @"task1"[K14] 3 retries, in <process "process"[K03]> (no vars)
R PROC_INST          COMPLETE       - #69->#66 K20 - BOUNDARY_EVENT "boundary" in <process "process"[K03]> !INVALID_STATE (Expected flow scope instance to be not interrupted but was interrupted by an event with id 'eventSP'.)
```

</p>
</details>

**Environment:**
- OS: Camunda Cloud
- Zeebe Version: 1.3.3 (probably also lower versions)

Causes #6957

### LINKED ISSUE #11121: ProcesExecutionRandomizedPropertyTest:shouldExecuteProcessToEnd is flaky
**Failures**
```
io.camunda.zeebe.test.util.stream.StreamWrapperException: No event found matching the criteria
	at io.camunda.zeebe.test.util.stream.StreamWrapper.await(StreamWrapper.java:72)
	at io.camunda.zeebe.engine.processing.randomized.ProcessExecutionRandomizedPropertyTest.shouldExecuteProcessToEnd(ProcessExecutionRandomizedPropertyTest.java:81)
	at jdk.internal.reflect.GeneratedMethodAccessor29.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:234)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:228)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:131)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
```
**Logs**

[Test results for Property Tests.zip](https://github.com/camunda/zeebe/files/10110849/Test.results.for.Property.Tests.zip)

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

Assume the following process, where we are waiting at `task`:
image

When we publish the `msgBoundary` message this will terminate the `task` and `subProcess`. This is done by sending at `TERMINATE` command for the `subProcess`. In order to activate the boundary event after termination an `EventTrigger` is created on the scope of this `subProcess`.

Before the `TERMINATE` command of the `subProcess` is processed, a message could be published to trigger the `eventSubProcess`. This will again result in a `TERMINATE` command for the `subProcess`. Again an `EventTrigger` is created, this time at the scope of the top-level process. Besides this the top-level process will be marked as interrupted, since this is an interrupting event sub process.

In this scenario one would expect that we don't take the boundary event. Instead we should terminate the `subProcess` and start processing the `eventSubProcess`. Because of the `EventTrigger` we have currently, this is not what happens. The boundary event gets activated and a `COMPLETE` command for this event is written. This command gets rejected, as the top-level process is in interrupted state. The top-level process never gets notified of the `subProcess` being terminated. Processing is stuck.

This PR adds a check to prevent activating the boundary event, by including a check if the flow scope of the `subProcess` is interrupted. If this is the case we terminate the process, and we will notify the top-level process that a child has been terminated.

## Related issues

closes #11121 
closes #8726

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
I have unrequested the review, there is more to this than I thought so the PR is incomplete for now.

**Comment by remcowesterhoud** [comment]
I should have all scenarios covered now

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/EmbeddedSubProcessTest.java [hunk 1]
```diff
@@ -564,4 +572,123 @@ public void shouldNotOverrideVariablesOnCompleteSubProcess() {
...
+
+    RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .withElementId("task")
+        .await();
+
+    // when
+    // We need to make sure no records are written in between the publish commands. This could
+    // cause the test to become flaky.
+    ENGINE.forEachPartition(ENGINE::pauseProcessing);
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/EmbeddedSubProcessTest.java:634]: :wrench: We don't need to pause the processing if we use `ENGINE.writeRecords()`. 

It writes the records in a batch and there will be no record in between.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/EmbeddedSubProcessTest.java [hunk 2]
```diff
@@ -564,4 +572,123 @@ public void shouldNotOverrideVariablesOnCompleteSubProcess() {
...
+            tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_TERMINATING),
+            tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_TERMINATED),
+            tuple(BpmnElementType.SUB_PROCESS, ProcessInstanceIntent.ELEMENT_TERMINATED),
+            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple(BpmnElementType.START_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            tuple(BpmnElementType.START_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATED))
+        .contains(
+            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED));
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/EmbeddedSubProcessTest.java:692]: :wrench: We could verify explicitly that the boundary event is not activated.

For example, by adding: `.doesNotContain(BpmnElementType.BOUNDARY_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATED)`.

**Review by saig0 — APPROVED** [review]
@remcowesterhoud looks great. :+1: 

I have two minor suggestions for the test. Please have a look.

**Comment by remcowesterhoud** [comment]
bors merge

**Pull request merged**